### PR TITLE
#8 [2/4]: Model fit helpers

### PR DESCRIPTION
**Pull request by zsusswein**
>[!IMPORTANT]
> This PR a trial for using [a stacked PR workflow](https://www.michaelagreiler.com/stacked-pull-requests/). Do not merge.

This PR builds on #14 to add two helper functions. Both are necessary prerequisites for the actual model fitting. They suggest a basis dimension for use in the model and construct the actual model formula.

The model formula constructor is meant to be easily extensible in the future to add hierarchy, day of week effects, and more. That isn't in scope for this PR, but if there are extensibility concerns it would be good to flag those now.

It would also be good to get feedback on some of the vibes-based decisions in here, The two big ones are the choice of basis dimension suggestion. It's a guess and I try to be honest in the documentation about that. I do think it's a _decent_ guess, but if there are real concerns that it's not great I can change things up. Likewise with the choice of three weeks per additional penalty basis. I think it's an approximately reasonable choice, but there will be tradeoffs. For each, I think it's important that we don't make the perfect the enemy of the good. Better to make some reasonable choices here to get to a v0.1 with a full package so we can do some more robust testing.

### CODE REVIEW

**Comment by zsusswein** [comment]
Opening against main to get codecov on CI, but will target #14 as the base for review.

**Review by seabbs — COMMENTED** [review]
This all looks reasonable and I like the majority of the docs a lot. 

I'd suggest another pas on the overall docs. I was also a little confused as to why the smoothness penalty was hard coded vs also being treated as user specified? 

I think ideally uses should be able to choose between adaptive or non-adaptive and I would just throw a warning if a adaptive spline is used with some small n. 

Before merging this PR I would look to see an issue investigating the smoothness defaults for the adaptive spline.

File: R/RtGam.R [hunk 1]
```diff
@@ -1,23 +1,90 @@
...
+#' degrees of freedom made available by `k` would be partitioned between the
+#' different smooths.
+#'
+#' In practice, GAMs penalize the wiggliness of smooth terms, so the fitted
+#' model will use fewer effective degrees of freedom than the total available.
+#' Although usually harmless to the model fit, excess degrees of freedom can
+#' make models slower to fit. However, models with `k` set too low may
+#' produce biased estimates or fail to converge.
+#'
+#' `RtGam` attempts to strike a reasonable balance between these concerns. It
```
[the remark this answers is not in the document]
Comment (seabbs) [reply R/RtGam.R:50]: you talk about this later elsewhere so perhaps just refactor here.

File: R/RtGam.R [hunk 2]
```diff
@@ -26,6 +93,95 @@ RtGam <- function(cases,
...
+#' needed. `k` should be increased to the maximum possible value. When running
+#' pre-set models in production, it would also be reasonable to fix the value
+#' of `k` above the default. Because GAMs penalize model wiggliness, the fit to
+#' both wiggly and non-wiggly data is likely to be satisfactory, at the cost of
+#' increased runtime.
+#'
+#' # Implementation details
+#' The algorithm to pick `k` is a piecewise function. When \eqn{n \le 10}, then
+#' the chosen value is \eqn{n}. When \eqn{n > 10}, then the selected value is
+#' \eqn{ \lceil \sqrt{10n} \rceil }.
```
[the remark this answers is not in the document]
Comment (zsusswein) [reply R/RtGam.R:152]: It is and I say that in the top-level para for this function:

https://github.com/CDCgov/cfa-gam-rt/blob/0b5cd8c9f8398c3ffa310baaa755e1c95bb18e28/R/RtGam.R#L115-L116

Are you suggesting that the default here should always be the dimension of the data? Or that I should clarify again here that `k` can be increased above this guess?

File: R/RtGam.R [hunk 3]
```diff
@@ -26,6 +93,95 @@ RtGam <- function(cases,
...
+dimensionality_heuristic <- function(n) {
+  # Input checks
+  rlang::check_required(n, "n", call = rlang::caller_env())
+  check_vector(n)
+  check_integer(n)
+  check_no_missingness(n)
+  check_elements_above_min(n, "n", min = 1)
+  check_vector_length(length(n), "n", min = 1, max = 1)
+
+  if (n < 10) {
```
[the remark this answers is not in the document]
Comment (zsusswein) [reply R/RtGam.R:182]: I'm not sure I'm following? The user is able to bypass this function to set `k` directly in `RtGam()`. Are you suggesting some additional argument to this helper function making the default guess change in some way?

**Comment by zsusswein** [comment]
> I was also a little confused as to why the smoothness penalty was hard coded vs also being treated as user specified?

This is good feedback. I think for the current functionality, the process is a little convoluted, But this is being developed with an eye to implementing day of week effects, partial pooling, etc. When implemented, we'll need to partition the degrees of freedom between these different components -- this process is how I propose to do that. We allow a user-specified global degrees of freedom and handle the partitioning between components internally. 

Perhaps we should allow the user to specify individual components degrees of freedom, but I think it's important we provide a decent first pass there by default. 

> I think ideally uses should be able to choose between adaptive or non-adaptive and I would just throw a warning if a adaptive spline is used with some small n.

This is good feedback. I'll implement here.

File: R/RtGam.R [hunk 4]
```diff
@@ -1,23 +1,90 @@
-#' Stub: Will become fitting/class constructor
+#' Fit a generalized additive model to incident cases
+#'
+#' # Model specification
+#' Incident cases are modeled as a smooth function of time with generalized
+#' additive models (GAMs). [RtGam] always fits a GAM, predicting incident
+#' cases as a smooth trend of time. However, the model adapts the penalty on
+#' wiggliness over time, allowing for changing epidemic dynamics.
```
Comment (athowes) [inline R/RtGam.R:7]: Is this only the case if there are >3 weeks of data? What about if there are <3 weeks of data?

File: R/RtGam.R [hunk 5]
```diff
@@ -26,6 +93,95 @@ RtGam <- function(cases,
...
+    is_grouped = !rlang::is_null(group)
+  )
 
   invisible(NULL)
 }
+
+#' Propose total basis dimensionality from number of data points
+#'
+#' Guess a reasonable value for the `k` argument of [RtGam] based on the number
+#' of data points. This guess **may not work** and almost certainly is not the
```
Comment (athowes) [inline R/RtGam.R:108]: "Return" rather than "Guess"

Comment (athowes) [reply R/RtGam.R:108]: What does it mean to "not work" -- would rephrase as "may be a poor choice"

File: R/RtGam.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (athowes) [reply R/RtGam.R:182]: Something like the function being `sqrt(10 * n))` and the user could set some aspect of that? e.g. `sqrt(x * n)` where `x` is an input.

If so, then we are providing less of a steer for the user. Of course it could be with a default `x = 10`.

**Review by athowes — APPROVED** [review]
Just getting up to speed with what's going on here so likely not very helpful review comments but in case helpful.

**Comment by athowes** [comment]
> I think ideally uses should be able to choose between adaptive or non-adaptive and I would just throw a warning if a adaptive spline is used with some small n.

I agree that the user should be able to choose between adaptive and non-adaptive. I guess that's not something we have implemented yet (as in so far there is just the one `k`, and the model isn't fit.)

Just looking, for the `group` option -- do we want to explicitly link that to "geographic". Is it not also possible to be other possible variables?

> When implemented, we'll need to partition the degrees of freedom between these different components -- this process is how I propose to do that. We allow a user-specified global degrees of freedom and handle the partitioning between components internally.

Feels like we might as well have the option for the user to specify the partition too, just the same default argument set-up as we have for the global (i.e. we have a heuristic function, and you can override it if you really want.)

File: R/RtGam.R [hunk 6]
```diff
@@ -1,23 +1,90 @@
...
+#'
+#' If more than three weeks of data are available, [RtGam] will fit a GAM with
+#' an adaptive spline basis. This basis is so named because it allows the
+#' wiggliness penalization to vary over time. Some parts of the fit can be more
+#' or less wiggly than other parts. If one part of the timeseries has a sudden
+#' change in trend while another part shows a smooth increase, the model can
+#' fit both components without smoothing away sharp changes or introducing
+#' additional artificial wiggliness.
+#'
+#' The model introduces an additional penalty basis dimension for each
```
[the remark this answers is not in the document]
Comment (zsusswein) [reply R/RtGam.R]: #19

File: R/RtGam.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (zsusswein) [reply R/RtGam.R]: See new changes and formatting.

File: R/RtGam.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (zsusswein) [reply R/RtGam.R]: Resolving per our discussion offline.

File: R/formula.R [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+#'  returned from [`prepare_inputs()`]
+#' @param k Global basis dimension to be partitioned between the model smooths
+#' @param is_grouped Whether to use a hierarchical model. Not yet supported.
+#' @return A formula to be used by [`mgcv::gam()`]
+#' @noRd
+formula_creator <- function(n_timesteps, k, is_grouped) {
+  outcome <- "cases"
+  intercept <- "1"
+
+  penalty_basis_dim <- penalty_basis_creator(n_timesteps)
```
[the remark this answers is not in the document]
Comment (zsusswein) [reply R/formula.R]: Refactored!

File: R/RtGam.R (hunk 4 above)
Comment (zsusswein) [reply R/RtGam.R]: Clarified in documentation update.

File: R/RtGam.R (hunk 5 above)
Comment (zsusswein) [reply R/RtGam.R]: Rewritten!

File: R/RtGam.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (zsusswein) [reply R/RtGam.R:169]: I'm not sure I love the idea of adding yet another knob to twist here. My philosophy on these helper functions is that they are meant to provide a steer. A sophisticated user can (and should!) modify `k` in the call to `RtGam()`.

Instead, I'd like to provide documentation on when and how `k` should be modified from the default and support the user in making that change directly. I've attempted to modify things in that direction with these revisions.

Please re-open if you have a problem with any of the above!

**Comment by zsusswein** [comment]
> This all looks reasonable and I like the majority of the docs a lot.

Did some rewriting and condensing 

> I agree that the user should be able to choose between adaptive and non-adaptive. I guess that's not something we have implemented yet (as in so far there is just the one k, and the model isn't fit.)

> I think ideally uses should be able to choose between adaptive or non-adaptive 

Done

>I would just throw a warning if a adaptive spline is used with some small n.

Will address in the next PR on the model fitting

> Before merging this PR I would look to see an issue investigating the smoothness defaults for the adaptive spline.

#19

>  for the group option -- do we want to explicitly link that to "geographic". Is it not also possible to be other possible variables?

Good catch. Modified it to be more general.

> Feels like we might as well have the option for the user to specify the partition too, just the same default argument set-up as we have for the global (i.e. we have a heuristic function, and you can override it if you really want.)

I want to keep this out of scope here, but will revisit once there's some actual partitioning happening (i.e., day of week effect and/or hierarchical model)

**Comment by zsusswein** [comment]
CI passing (except NEWS.md)

File: R/formula.R [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+#' @param is_grouped Whether to use a hierarchical model. Not yet supported.
+#' @return A formula to be used by [`mgcv::gam()`]
+#' @noRd
+formula_creator <- function(k, m, is_grouped) {
+  outcome <- "cases"
+  intercept <- "1"
+
+  smooth_basis_dim <- smooth_basis_creator(k)
+
+  # Apply adaptive spline if 3 weeks or more of data are available
```
Comment (seabbs) [inline R/formula.R:29]: This comment seems like it could be easy to let get out of date. Why not just doc where m is set

File: R/formula.R [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    # order of the smoothing penalty as it does in the other smoothing bases.
+    plus_global_trend <- glue::glue("+ s(timesteps,
+                                         k = {smooth_basis_dim[['global_trend']]},
+                                         m = {m},
+                                         bs = 'ad')") # nolint
+  } else {
+    # Adaptive penalty with `m = 1` is equivalent to a non-adaptive smooth but
+    # thin-plate performance is a bit better than p-spline, so preference to
+    # fall back to thin-plate.
+    plus_global_trend <- glue::glue("+ s(timesteps,
```
Comment (seabbs) [inline R/formula.R:42]: This is a very boring comment. 

It might read more naturally if the plus here was in line 48. This would also slightly reduce copy and paste.

**Review by seabbs — APPROVED** [review]
Overall looks good to me. A very few very minor additional comments.

File: R/formula.R (hunk 9 above)
Comment (zsusswein) [reply R/formula.R:42]: Ah yes, I thought about this when I was designing this. The idea is that we want to implement optional local trends and day of week effects. So the formula will be something like

`{outcome} ~ {intercept} {plus_global_trend} {plus_local_trend} {plus_day_of_week}`

If instead it were

`{outcome} ~ {intercept} + {global_trend} + {local_trend} + {day_of_week}` 

and we take out the local trend or day of week then the dangling `+` would cause the formula parsing to error. By assigning the `+` in the variable to be evaluated we can mix and match as needed.

Comment (seabbs) [reply R/formula.R:42]: I see. that is quite a compelling arg. My counter arg is can you just on the fly construct a string with + between the parts? that seems like the better option.

Comment (seabbs) [reply R/formula.R:42]: Above would reduce the + spam quite a bit once there are multiple formula components in play

Comment (zsusswein) [reply R/formula.R:42]: You mean like this?

```r
global_trend <- "s(x)"
local_trend <- "s(y)"
day_of_week <- NULL

formula <- paste("outcome ~ 1", global_trend, local_trend, day_of_week, sep = "+")
```

That still doesn't work because the last NULL is still evaluated so it returns

>  "outcome ~ 1+s(x)+s(y)+"

Comment (seabbs) [reply R/formula.R:42]: I did mean like this but I hadn't thought about the null handling. You could sidestep it by first composing what will be combined (and drop NULLS) but it starts to get less clear. I think leave as is.
